evmrpc/simulate.go

```diff
@@ -295,16 +295,11 @@ func (b *Backend) StateAtTransaction(ctx context.Context, block *ethtypes.Block,
 		// set block context time as of the block time (block time is the time of the CURRENT block)
 		blockContext.Time = block.Time()
 
-		// set address association for the sender if not present. Note that here we take the shortcut
-		// of querying from the latest height with the assumption that if this tx has been processed
-		// at all then its association must be present in the latest height
-		_, associated := b.keeper.GetSeiAddress(statedb.Ctx(), msg.From)
-		if !associated {
-			seiAddr, associatedNow := b.keeper.GetSeiAddress(b.ctxProvider(LatestCtxHeight), msg.From)
-			if !associatedNow {
-				return nil, vm.BlockContext{}, nil, nil, fmt.Errorf("address %s is not associated in the latest height", msg.From.Hex())
-			}
-			if err := ante.NewEVMPreprocessDecorator(b.keeper, b.keeper.AccountKeeper()).AssociateAddresses(statedb.Ctx(), seiAddr, msg.From, nil); err != nil {
+		// check if this transaction caused legacy address association and simulate if so
+		addressThen := b.keeper.GetSeiAddress(statedb.Ctx(), msg.From)
+		addressNow := b.keeper.GetSeiAddress(b.ctxProvider(LatestCtxHeight), msg.From)
+		if !addressThen.Equals(addressNow) {
+			if err := ante.NewEVMPreprocessDecorator(b.keeper, b.keeper.AccountKeeper()).AssociateAddresses(statedb.Ctx(), addressNow, msg.From, nil); err != nil {
```

Comment:
see above. An example would be:
- tx T created a legacy association for account 0xXYZ at height 100
- latest height is 200
- when we trace T, we start with the state at height 99
- `addressThen` here would be the association of 0xXYZ at height 99, which is the direct cast because T hasn't materialized yet
- `addressNow` here would be the association at height 200, which would be the legacy association as it's set on height 100
- since those two are different, we know an address association is set on height 100 (and only a legacy association can cause the difference), so we simulate it here

Reply:
ty for explanation